runtime/onert/core/src/util/EventRecorder.h

```diff
@@ -17,28 +17,75 @@
 #ifndef __ONERT_UTIL_EVENT_RECORDER_H__
 #define __ONERT_UTIL_EVENT_RECORDER_H__
 
+#include "util/TracingCtx.h"
+
 #include <map>
 #include <memory>
 #include <mutex>
 
 #include <vector>
 
+struct SubgDurationEvent;
+struct OpDurationEvent;
+
+class DurationEventVisitor
+{
+public:
+  virtual ~DurationEventVisitor() = default;
+
+  virtual std::string visit(const SubgDurationEvent &) const
+  {
+    throw std::runtime_error("Please implement");
+  }
+  virtual std::string visit(const OpDurationEvent &) const
+  {
+    throw std::runtime_error("Please implement");
+  }
+};
+
+// refer to https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit#
 struct Event
 {
-  std::string name;
-  std::string tid;
-  std::string ph;                                        /* REQUIRED */
-  std::string ts;                                        /* REQUIRED */
+  const onert::util::TracingCtx *tracing_ctx;
+
+  std::string ph;                                        // Event type.
+  std::string ts;                                        // tracing clock of timestamp of this event
   std::vector<std::pair<std::string, std::string>> args; // user-defined data: pairs of (key, value)
+
+  virtual ~Event() = default;
 };
 
 struct DurationEvent : public Event
 {
-  // TO BE FILLED
+  uint32_t session_index;
+  uint32_t subg_index;
+
+  virtual std::string accept(DurationEventVisitor &visitor) const = 0;
+
+protected:
+  DurationEvent() = default;
+};
+
+struct SubgDurationEvent : public DurationEvent
+{
+  std::string accept(DurationEventVisitor &visitor) const override { return visitor.visit(*this); }
+};
+
+struct OpDurationEvent : public DurationEvent
+{
+  // Note: DurationEvent's name and tid will be set by EventWriter
+  std::string backend;
+  uint32_t op_index;
+  std::string op_name;
+  uint32_t op_seq_size; // if this event is for an operation sequence of multiple operations
+
+  std::string accept(DurationEventVisitor &visitor) const override { return visitor.visit(*this); }
 };
 
```

Comment:
With this PR, I tried NOT to touch any memory measuring code including `CounterEvent`. So it has almost same member variables as before.